cli/popper/runner_slurm.py

```diff
@@ -1,249 +1,144 @@
 
-        # start a tail process on the output and error file
-        self.start_output_error_stream(out_file, err_file)
+        # start a tail (background) process on the output file
+        self._start_out_stream(out_file)
 
-        # submit the job and wait, then parse the job_id
-        ecode, output = pu.exec_cmd(sbatch_cmd.split(" "), logging=False)
-        job_id = int(output.split(" ")[-1].strip("\n"))
+        # submit the job and wait
+        ecode, output = HostRunner._exec_cmd(sbatch_cmd, logging=False)
 
         # kill the tail process
-        self.stop_output_error_stream()
+        self._stop_output_error_stream()
+
+        self._spawned_jobs.remove(job_name)
 
-        SlurmRunner.spawned_jobs.remove(job_name)
         return ecode
 
-    @staticmethod
-    def cancel_job():
-        for job_name in SlurmRunner.spawned_jobs:
+    def stop_running_tasks(self):
+        for job_name in self._spawned_jobs:
             log.info(f'Cancelling job {job_name}')
-            ecode, _ = pu.exec_cmd(["scancel", "--name", job_name])
-            if ecode:
-                log.fail(f"Failed to cancel the job {job_name}.")
+            _, ecode, _ = HostRunner._exec_cmd(['scancel', '--name', job_name])
+            if ecode != 0:
+                log.warning(f'Failed to cancel the job {job_name}.')
 
 
-class DockerRunner(SlurmRunner):
-    spawned_containers = set()
+class DockerRunner(SlurmRunner, HostDockerRunner):
 
-    def __init__(self, config):
-        super(DockerRunner, self).__init__(config)
+    def __init__(self, **kw):
+        super(DockerRunner, self).__init__(init_docker_client=False, **kw)
 
     def __exit__(self, exc_type, exc, traceback):
-        spawned_containers = set()
+        pass
 
     def run(self, step):
-        """Execute the given step in docker."""
-        # generate cid
-        cid = pu.sanitized_name(step['name'], self.config.wid)
-        step['cmd_list'] = []
+        """Execute the given step via slurm in the docker engine."""
+        cid = pu.sanitized_name(step['name'], self._config.wid)
+        cmd = []
 
-        # prepare image build artifacts
-        build, img, dockerfile = HostDockerRunner.get_build_info(
-            step, self.config.workspace_dir, self.config.workspace_sha)
+        build, img, tag, dockerfile = self._get_build_info(step)
 
-        if build:
-            DockerRunner.docker_build(
-                step, img, dockerfile, self.config.dry_run)
-        elif not self.config.skip_pull and not step.get('skip_pull', False):
-            DockerRunner.docker_pull(step, img, self.config.dry_run)
+        cmd.append(f'docker rm -f {cid} || true')
 
-        # remove container if it exists
-        DockerRunner.docker_rm(step, cid, self.config.dry_run)
+        if build:
+            cmd.append(f'docker build {img}:{tag} {dockerfile}')
+        elif not self._config.skip_pull and not step.get('skip_pull', False):
+            cmd.append(f'docker pull {img}:{tag}')
```

Comment:
Can we redirect the output of these build/pull commands to `/dev/null`, otherwise they would show up in the output/error stream ?

Reply:
that's the goal, to show the output of this operations, unless `--quiet` is given. But now that you mention it, we should be using `log.step_info` instead of `info()` from the `HostRunner._exec_cmd()` function. I'll change it

Reply:
Yeah. Thanks